gsplat/rendering.py

```diff
@@ -213,6 +232,30 @@ def rasterization(
             colors.dim() == 4 and colors.shape[:2] == (C, N) and colors.shape[3] == 3
         ), colors.shape
         assert (sh_degree + 1) ** 2 <= colors.shape[-2], colors.shape
+        if distributed:
+            assert (
+                colors.dim() == 3
+            ), "Distributed mode only supports per-Gaussian colors."
+
+    # If in distributed mode, we distribute the projection computation over Gaussians
+    # and the rasterize computation over cameras. So first we gather the cameras
+    # from all ranks for projection.
+    if distributed:
+        world_rank = torch.distributed.get_rank()
+        world_size = torch.distributed.get_world_size()
+        N_world = [None] * world_size
+        C_world = [None] * world_size
+        torch.distributed.all_gather_object(N_world, N)
+        torch.distributed.all_gather_object(C_world, C)
+
+        # [TODO] `all_gather` is not differentiable w.r.t. viewmats and Ks
+        out_tensor_list = [torch.empty((C_i, 4, 4), device=device) for C_i in C_world]
+        torch.distributed.all_gather(out_tensor_list, viewmats.contiguous())
+        viewmats = torch.cat(out_tensor_list, dim=0)
+        out_tensor_list = [torch.empty((C_i, 3, 3), device=device) for C_i in C_world]
+        torch.distributed.all_gather(out_tensor_list, Ks.contiguous())
+        Ks = torch.cat(out_tensor_list, dim=0)
+        C = len(viewmats)  # Silently change C from local #Cameras to global #Cameras.
```

Comment:
@TarzanZhao  The current design of this API is that it takes in the local cameras that needs to be rendered by this rank. So before the projection stage, an all_gather needs to be done for each rank to get access to all cameras. 

I guess another option for the API design is that this function takes in the "global" cameras (all cameras needed to be rendered by all ranks), which would avoid this all_gather operation. But this would mean that user have to load the same global set of cameras in each rank, but only supervise on a subset of them, which feels a bit counter intuitive from the user experience perspective.

I personally prefer the first design choice but that means two all_gather and sync operations here. Do you think the affect of this would be large in your experience?

Reply:
Using all_gather_object() is often slow in my experience because it transfers data from the CPU to the GPU, then to another GPU, and finally back to the CPU. It's advisable to avoid using this function as much as possible. If you prefer the first design, then please try to avoid using all_gather_object() at least. 

In our Grendel's implementation, every gpu keep the same dataset and uses the same random seed to generate the same batch across GPU every time. This is the other option that you have mentioned. Then there is no need for all-gather cameras.